File: 2888-minimum-index-of-a-valid-split/2888-minimum-index-of-a-valid-split.py
class Solution:
    def minimumIndex(self, nums: List[int]) -> int:
        # Step 1: Boyer–Moore to find dominant candidate
        dom_ele = nums[0]
        counter = 1

        for i in range(1, len(nums)):
            if counter == 0:
                dom_ele = nums[i]
                counter = 1
            elif nums[i] == dom_ele:
                counter += 1
            else:
                counter -= 1 

        # Step 2: Confirm if it's actually dominant
        temp_cnt = nums.count(dom_ele)
        if temp_cnt * 2 <= len(nums):  # Must be strictly > n/2
            return -1

        # Step 3: Check valid split point
        c = 0  # Count of dom_ele in prefix
        for i in range(len(nums) - 1):
            if nums[i] == dom_ele:
                c += 1
            if (c > (i + 1) // 2 and (temp_cnt - c) > (len(nums) - i - 1) // 2):
                return i

        return -1

File: 2888-minimum-index-of-a-valid-split/2888-minimum-index-of-a-valid-split.java
class Solution {
    public int minimumIndex(List<Integer> nums) {
        //Find a candidate for the dominant element using Boyer–Moore Voting Algorithm.
        int ele = nums.get(0); // Candidate element initialized to the first element.
        int k = 1;             // Counter for the candidate.
        for (int x : nums) {
            if (k == 0) {
                // Reset the candidate when the counter is zero.
                ele = x;
                k = 1;
            } else if (x != ele) {
                // Decrement the counter when the current element does not match the candidate.
                k--;
            } else {
                // Increment the counter when the current element matches the candidate.
                k++;
            }
        }
        
        // Step 2
        k = 0;
        for (int x : nums) {
            if (x == ele)
                k++; 
        }
        // If the candidate doesn't appear more than half the time, return -1.
        if (k < nums.size() / 2)
            return -1;

        // Step 3: Find the minimum valid split index.
        int presum = 0; 
        int ans = -1;   
        for (int i = 0; i < nums.size(); i++) {
            if (nums.get(i) == ele) {
                presum++; 
            }
            // Left subarray (nums[0...i]): candidate count > (i+1)/2.
            // Right subarray (nums[i+1...end]): candidate count > (nums.size() - i - 1)/2.
            if (presum > (i + 1) / 2 && k - presum > (nums.size() - i - 1) / 2) {
                ans = i; // Valid split found.
                break;
            }
        }
        return ans;
    }
}
